modules/tensor_mechanics/src/kernels/TotalLagrangianStressDivergence.C

```diff
@@ -0,0 +1,209 @@
+    }
+  }
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpResidual()
+{
+  return _pk1[_qp].row(_component) * _grad_test[_i][_qp];
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpJacobian()
+{
+  Real value = 0.0;
+  value += materialJacobian(_component, _component, _grad_test[_i][_qp], _grad_phi[_j][_qp]);
+
+  return value;
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  Real value = 0.0;
+
+  for (unsigned int cc = 0; cc < _ndisp; cc++)
+  {
+    if (jvar == _disp_nums[cc])
+    {
+      value +=
+          materialJacobian(_component, cc, _grad_test[_i][_qp], _disp_vars[cc]->gradPhi()[_j][_qp]);
+      break;
+    }
+  }
+
+  return value;
+}
+
+Real
+TotalLagrangianStressDivergence::materialJacobian(unsigned int i,
+                                                  unsigned int k,
+                                                  const RealGradient & grad_phi,
+                                                  const RealGradient & grad_psi)
+{
+  Real value = 0.0;
+  for (unsigned int j = 0; j < _ndisp; j++)
+    for (unsigned int l = 0; l < _ndisp; l++)
+      value += _dpk1[_qp](i, j, k, l) * grad_phi(j) * grad_psi(l);
+
+  return value;
+}
+
+void
+TotalLagrangianStressDivergence::computeOffDiagJacobianScalar(unsigned int jvar)
+{
+  if (jvar == _macro_gradient_num)
+  {
+    DenseMatrix<Number> & ken = _assembly.jacobianBlock(_var.number(), jvar);
+    DenseMatrix<Number> & kne = _assembly.jacobianBlock(jvar, _var.number());
+
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      Real dV = _JxW[_qp] * _coord[_qp];
+      for (_h = 0; _h < _macro_gradient->order(); _h++)
+      {
+        for (_i = 0; _i < _test.size(); _i++)
+        { // This assumes Galerkin, i.e. the test and trial functions are the
+          // same
+          ken(_i, _h) += computeBaseJacobian() * dV;
+          kne(_h, _i) += computeConstraintJacobian() * dV;
+        }
+      }
+    }
+  }
+}
+
+Real
+TotalLagrangianStressDivergence::computeBaseJacobian()
+{
+  return materialBaseJacobian();
+}
+
+Real
+TotalLagrangianStressDivergence::materialBaseJacobian()
+{
+  Real value = 0.0;
+  for (unsigned int j = 0; j < _ndisp; j++)
+    value +=
+        _dpk1[_qp](_component, j, _indices[_h].first, _indices[_h].second) * _grad_test[_i][_qp](j);
+
+  return value;
+}
+
+Real
+TotalLagrangianStressDivergence::computeConstraintJacobian()
+{
+  if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Stress)
+  {
+    return materialConstraintJacobianStress();
+  }
+  else if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Strain)
+  {
+    if (_ld)
+      return ldConstraintJacobianStrain();
+    else
+      return sdConstraintJacobianStrain();
+  }
+  else
+  {
+    mooseError("Unknown constraint type in kernel calculation!");
+  }
+}
+
+Real
+TotalLagrangianStressDivergence::sdConstraintJacobianStrain()
```

Comment:
I'll say it here: I think these names might be misleading. These names are almost suggesting that we are computing some sort of stress/strain. Can we think of a more concise name?

Reply:
This is the "small deformation Jacobian contribution for a strain constraint."  The name makes sense to me.  Given that this is what it does (and the others are similar) would something else be more clear?